winrm/winrm.go

```diff
@@ -0,0 +1,235 @@
+// copyright 2016 canonical ltd.
+// copyright 2016 cloudbase solutions
+// licensed under the lgplv3, see licence file for details.
+
+package winrm
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/masterzen/winrm"
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+const (
+	// default http,https winrm ports
+	httpPort            = 5985
+	httpsPort           = 5986
+	defaultWinndowsUser = "Administrator"
+)
+
+var (
+	logger = loggo.GetLogger("juju.utils.winrm")
+)
+
+// Client type retains information about the winrm connection
+type Client struct {
+	conn   *winrm.Client
+	pass   string
+	secure bool
+}
+
+// Password returns the winrm connection password to be used
+// to enable certification inseriont into the cert store location
+// If the default client of this pkg is not set this will panic.
+func (c Client) Password() string {
+	return c.pass
+}
+
+// Secure returns true if the client is using a secure connection or false
+// if it's just a normal http
+func (c Client) Secure() bool {
+	return c.secure
+}
+
+// ClientConfig used for setting up a secure https client connection
+type ClientConfig struct {
+	// User of the client connection
+	// If you want the default user Administrator, leave this empty
+	User string
+	// Host where we want to connect
+	Host string
+	// Key Private RSA key
+	Key []byte
+	// Cert https client x509 cert
+	Cert []byte
+	// CACert of the host we wish to connect
+	// This can be nil if Insecure is false
+	CACert []byte
+	// Timeout on how long we should wait until a connection is made.
+	// If empty this will use the 60*time.Seconds winrm default value
+	Timeout time.Duration
+	// Insecure flag for checking the CACert or not. If this is true there should
+	// be a valid CACert passed also.
+	Insecure bool
+	// Password callback for returning a password in different ways
+	Password GetPasswd
+}
+
+// errNoPasswdFn sentinel for returning unset password callback
+var errNoPasswdFn = fmt.Errorf("No password callback")
+
+// password checks if the password callback is set and returns the password
+// from that specific get password handler
+func (c ClientConfig) password() (string, error) {
+	if c.Password != nil {
+		// make it look like a dropdown shell
+		fmt.Fprintf(os.Stdout, "[Winrm] >> ")
```

Comment:
please move the fmt.Fprintf prompts to the GetPasswd implementation. this code shouldn't assume GetPasswd is interactive.